python/pyspark/sql/tests.py

```diff
@@ -2620,6 +2635,178 @@ def range_frame_match():
 
         importlib.reload(window)
 
+
+class TypesTest(unittest.TestCase):
```

Comment:
Many codes, in particular, tests were ported from https://github.com/apache/spark/pull/17227. Many efforts were put here. I think this PR should credit to @dgingrich if there are not many things that should be changed.